pip install annoy not working on macOS mojave

This is where `pip install annoy` fails:

```bash
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/taegyung/miniconda3/include -arch x86_64 -I/Users/taegyung/miniconda3/include -arch x86_64 -I/Users/taegyung/miniconda3/include/python3.6m -c src/annoymodule.cc -o build/temp.macosx-10.7-x86_64-3.6/src/annoymodule.o -O3 -ffast-math -fno-associative-math -march=native
  warning: include path for stdlibc++ headers not found; pass '-std=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
  In file included from src/annoymodule.cc:15:
  src/annoylib.h:49:10: fatal error: 'vector' file not found
  #include <vector>
           ^~~~~~~~
  1 warning and 1 error generated.
  error: command 'gcc' failed with exit status 1
```

I tried some things I found from previous issues and some general solutions accros the web, including

- sudo pip install annoy
- Installing XCode 9.4 (I wasn't using any version of XCode previously)
- xcode-select --install
- brew remove gcc; brew install gcc

I finally got it by running `conda install -c akode annoy`, but I'm not sure if this is the right way to do it (and don't know why this even works.)

--- a/setup.py
+++ b/setup.py
@@ -18,6 +18,7 @@
 from setuptools import setup, Extension
 import codecs
 import os
+import platform
 import sys
 
 readme_note = """\
@@ -34,23 +35,26 @@ readme_note = """\
 with codecs.open('README.rst', encoding='utf-8') as fobj:
     long_description = readme_note + fobj.read()
 
+# Various platform-dependent extras
+extra_compile_args = []
+extra_link_args = []
+
 if os.environ.get('TRAVIS') == 'true':
     # Resolving some annoying issue
-    travis_extra_compile_args = ['-mno-avx']
-else:
-    travis_extra_compile_args = []
+    extra_compile_args += ['-mno-avx']
 
 # Not all CPUs have march as a tuning parameter
-import platform
 cputune = ['-march=native',]
-if platform.machine() == "ppc64le":
-    cputune = ['-mcpu=native',]
+if platform.machine() == 'ppc64le':
+    extra_compile_args += ['-mcpu=native',]
 
 if os.name != 'nt':
-    compile_args = ['-O3', '-ffast-math', '-fno-associative-math']
-else:
-    compile_args = []
-    cputune = []
+    extra_compile_args += ['-O3', '-ffast-math', '-fno-associative-math']
+
+# #349: something with OS X Mojave causes libstd not to be found
+if platform.system() == 'Darwin':
+    extra_compile_args += ['-std=c++11', '-mmacosx-version-min=10.9']
+    extra_link_args += ['-stdlib=libc++', '-mmacosx-version-min=10.9']
 
 setup(name='annoy',
       version='1.15.0',
@@ -60,7 +64,8 @@ setup(name='annoy',
         Extension(
             'annoy.annoylib', ['src/annoymodule.cc'],
             depends=['src/annoylib.h', 'src/kissrandom.h', 'src/mman.h'],
-            extra_compile_args=compile_args + cputune + travis_extra_compile_args
+            extra_compile_args=extra_compile_args,
+            extra_link_args=extra_link_args,
         )
       ],
       long_description=long_description,

